### PR TITLE
Bug 2033065: fix pod log height issue

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Issue
Addresses https://bugzilla.redhat.com/show_bug.cgi?id=2033065 / https://issues.redhat.com/browse/OCPBUGSM-38515
Also organizes the LogHeader in an OverflowMenu

# Description
Log viewer toolbar does not wrap correctly. The height of the LogView is always set to `100%`. This causes the header elements to align undesirably and the logs are not readable on mobile screen.
Screenshot 2022-04-19 at 7 11 33 PM

# Fix Screenshots 
------------------------
ezgif com-gif-maker (2)

# Tests
none

# Browser conformance
Chrome

/cherrypick release-4.9
/cherrypick release-4.10

### CODE REVIEW

**Comment by spadgett** [comment]
Does PF have a collapsible toolbar where there is an overflow menu for small screens?

**Comment by abhinandan13jan** [comment]
@spadgett Yes, there is an `collapseListedFilters` breakpoint prop available

**Comment by abhinandan13jan** [comment]
Should we try a collapsible here? @spadgett

**Comment by abhinandan13jan** [comment]
@spadgett This is how it looks in the OverflowMenu. Please let me know whether to proceed with this?
Screenshot 2022-05-05 at 7 04 53 PM
Screenshot 2022-05-05 at 7 05 04 PM

**Comment by spadgett** [comment]
@abhinandan13jan Thank you, this looks like a better approach to me 👍 

Adding @megan-hall for UX input on the items and position of the overflow menu.

**Comment by megan-hall** [comment]
> @abhinandan13jan Thank you, this looks like a better approach to me 👍
> 
> Adding @megan-hall for UX input on the items and position of the overflow menu.

I wonder if we should keep the overflow menu to the right of search - just for consistency since that is where the actions appear when the window is expanded?

**Comment by megan-hall** [comment]
Is something like this possible
Screen Shot 2022-05-05 at 4 59 43 PM
?

**Comment by abhinandan13jan** [comment]
Hello @megan-hall, there are two issues around the design requested. 
**The `overflowMenu` doesn't have a placement property and will always open to its right unless we do a CSS-hack. So, it cannot be placed at the extreme right.**

Screenshot 2022-05-06 at 4 58 43 PM

**The search widget expands on input and it is not ideal to fit it in a row of three elements.**

Screenshot 2022-05-06 at 4 58 53 PM

--------------------------------------

### We can implement either of the following. 
**1. Add the dropdown collapse icon in the middle(end of row) the first row.**
Screenshot 2022-05-06 at 5 23 42 PM

Screenshot 2022-05-06 at 5 23 32 PM


**2. Rearrange and have it on the right of second row**
Screenshot 2022-05-06 at 5 37 20 PM


**3. Put it on the left of either of the rows.**


Please left me know how to proceed? cc: @spadgett @invincibleJai

**Comment by abhinandan13jan** [comment]
Raised PF Bug https://github.com/patternfly/patternfly-react/issues/7398

**Comment by abhinandan13jan** [comment]
Updated PR to use option 2. Currently looks like the following @megan-hall 
Screenshot 2022-05-11 at 6 29 25 PM

**Comment by spadgett** [comment]
/assign @TheRealJon 

Jon, can you help review since you're familiar with the log components?

**Review by TheRealJon — CHANGES_REQUESTED** [review]
See https://github.com/abhinandan13jan/console/pull/6

**Comment by abhinandan13jan** [comment]
Thanks @TheRealJon for the suggestions. I've updated the PR and tested locally

**Review by TheRealJon — CHANGES_REQUESTED** [review]
I did some testing on this and found a few issues. It seems like there's been a lot of back and forth on the design of this, but the responsive behavior of this solution is a little weird. Here's a screen capture of the testing I did:
screencast-meet google com-2022 06 01-14_23_18

- The actions are no longer right-aligned. 
- They begin wrapping right before the breakpoint where they collapse into an overflow menu. We probably need to use a '2xl' breakpoint to make these actions collapse much sooner. 
- The search bar wraps to a second line at smaller widths, which takes up valuable log viewer screen space and just doesn't quite look right. Even though it's an improvement over the previous behavior, IMO, it could be even better.
 
I'm wondering if this is an opportunity to use the PatternFly `Toolbar` and specifically the `ToolbarToggleGroup` component. Has that come up as an option during previous discussions? See an example here: https://www.patternfly.org/v4/components/toolbar#with-filters

This would allow us to collapse the search, container dropdown, and current/previous log dropdown at smaller widths:

screencast-github com-2022 06 01-14_46_52

File: frontend/public/components/utils/resource-log.tsx [hunk 1]
```diff
@@ -148,6 +159,45 @@ const showDebugAction = (pod: PodKind, containerName: string) => {
   return !containerStatus?.state?.running || !containerStatus.ready;
 };
 
+const getPodLogLinkUrl = (podLogLink, templateValues) => {
```
Comment (jerolimov) [inline frontend/public/components/utils/resource-log.tsx:162]: Can you add some times here please.

File: frontend/public/components/utils/resource-log.tsx [hunk 2]
```diff
@@ -148,6 +159,45 @@ const showDebugAction = (pod: PodKind, containerName: string) => {
...
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.warn('invalid log link regex', namespaceFilter, e);
+      return '';
+    }
+  }
+  return replaceVariables(hrefTemplate, templateValues);
+};
+
+const getPodLogExternalLinks = (podLogLinks, templateValues) =>
```
Comment (jerolimov) [inline frontend/public/components/utils/resource-log.tsx:179]: Can you also add some times here please. Thanks

File: frontend/public/components/utils/resource-log.tsx (hunk 1 above)
Comment (abhinandan13jan) [reply frontend/public/components/utils/resource-log.tsx]: updated

File: frontend/public/components/utils/resource-log.tsx (hunk 2 above)
Comment (abhinandan13jan) [reply frontend/public/components/utils/resource-log.tsx]: updated

File: frontend/public/components/utils/resource-log.tsx [hunk 3]
```diff
@@ -239,124 +299,137 @@ export const LogControls: React.FC<LogControlsProps> = ({
...
+                onFocus={() => {
+                  if (status === STREAM_ACTIVE) {
+                    toggleStreaming();
+                  }
+                }}
+                placeholder="Search"
+              />
+            </OverflowMenuItem>
+          </OverflowMenuGroup>
+          <OverflowMenuGroup className="co-toolbar__group co-toolbar__group--right">
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/public/components/utils/resource-log.tsx]: changed

File: frontend/public/components/utils/resource-log.tsx [hunk 4]
```diff
@@ -368,6 +441,15 @@ export const ResourceLog: React.FC<ResourceLogProps> = ({
...
 }) => {
   const { t } = useTranslation();
+  const [height, setHeight] = React.useState<string>(LOG_FULLSCREEN_HEIGHT);
+  const isMobile = useIsMobile();
+
+  if (isMobile) {
+    height !== LOG_MOBILE_HEIGHT && setHeight(LOG_MOBILE_HEIGHT);
+  } else {
+    height !== LOG_FULLSCREEN_HEIGHT && setHeight(LOG_FULLSCREEN_HEIGHT);
+  }
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/public/components/utils/resource-log.tsx:443]: I understand. But maybe this is the best bet we have right now as the height if not explicitly specified the content issue remains. Hence it was discussed with UX for the same. I've changed the height to 300px. Do let me know what is to be done here?

Without specific height:
Screenshot 2022-06-16 at 9 07 25 PM

File: frontend/public/components/utils/resource-log.tsx [hunk 5]
```diff
@@ -239,124 +299,137 @@ export const LogControls: React.FC<LogControlsProps> = ({
...
+  });
+
+  const overflowActions = [
+    logURLLink,
+    downloadLink,
+    wrapLoglinesCheckbox,
+    ...(shouldShowDebugAction ? [debugAction] : []),
+    ...podLogExternalLinks,
+    ...(screenfull.enabled ? [fullscreenButton] : []),
+  ];
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/public/components/utils/resource-log.tsx]: updated, now it looks like the description screenshot

**Comment by megan-hall** [comment]
LGTM! Thanks for all the collaboration on this!

**Review by TheRealJon — CHANGES_REQUESTED** [review]
I've talked with @megan-hall about this and we came to a consensus that we should just try to match the behaviors of the [PatternFly toolbar example](https://www.patternfly.org/v4/components/toolbar#with-filters). So, the "actions" (view raw, download, toggle wrap lines, etc.) should collapse into a kebab dropdown at the 2xl breakpoint, then the search, container, and log type dropdowns should collapse into a "filter" dropdown at the xl breakpoint. Here is a rough example that I worked up using the ToolbarToggleGroup and OverflowMenu components. This is not fully hashed out or tested, but it should get you most of the way there:

```tsx
    <Toolbar>
      <ToolbarContent>
        <ToolbarItem>{showStatus()}</ToolbarItem>
        <ToolbarToggleGroup toggleIcon={<FilterIcon />} breakpoint="lg">
          <ToolbarItem variant="search-filter">
            <LogViewerSearch
              onFocus={() => {
                if (status === STREAM_ACTIVE) {
                  toggleStreaming();
                }
              }}
              placeholder="Search"
            />
          </ToolbarItem>
          {dropdown && <ToolbarItem>{dropdown}</ToolbarItem>}
          <ToolbarItem>{logTypeSelect(!hasPreviousLog)}</ToolbarItem>
        </ToolbarToggleGroup>
        <ToolbarItem>
          <OverflowMenu breakpoint="2xl">
            <OverflowMenuContent>
              {logActions.map((action, index) => {
                return <ToolbarItem key={`log-control-item-${index}`}>{action}</ToolbarItem>;
              })}
            </OverflowMenuContent>
            <OverflowMenuControl>
              <Dropdown
                toggle={<KebabToggle onToggle={onToggleOverflowDropdown} />}
                isOpen={isOverflowDropdownOpen}
                isPlain
                dropdownItems={logActions.map((action, index) => {
                  return (
                    <DropdownItem key={`log-control-overflow-item-${index}`}>{action}</DropdownItem>
                  );
                })}
              />
            </OverflowMenuControl>
          </OverflowMenu>
        </ToolbarItem>
      </ToolbarContent>
    </Toolbar>
```

**Comment by TheRealJon** [comment]
I opened a PatternFly issue for the ToolbarToggleGroup component to address the problem you are seeing where you can't use multiple ToolbarToggleGroups in a single Toolbar: https://github.com/patternfly/patternfly-react/issues/7590

**Comment by jerolimov** [comment]
Hey @abhinandan13jan, on the desktop this PR changes the order of the input fields, that's maybe fine.

### 1. But couldn't we span and right align the last group (Wrap lines ... Expand) ??

Before:
image

Now:
image

Since this is a flex layout this should be possible with `justify-content: flex-end`:
image

For sure, this must be added to a specific breakpoint, but the different Toolbar components has props for that, or?

### 2. When reducing the window size to a "medium width" I got a more action (3 dots) which jumps around when clicking it:

https://user-images.githubusercontent.com/139310/175024016-adfe9a99-69e7-4867-a2de-abb7691a0a7b.mp4

### 3. And then I'm personally also confused when reducing the width, even more, why do we have a filter and a more (3 dots) button which does the same?

https://user-images.githubusercontent.com/139310/175024129-929f3205-ef21-4682-bf71-c4ba6f9af701.mp4

### 4. And finally, there is a fix height, as someone mentioned above. It makes sense for me to have a min-height to force a scrollbar, but a fix or max-height doesn't allow the user to use the full height of his phone.

https://user-images.githubusercontent.com/139310/175025061-a5096d57-328f-4c44-b89f-ea3cfab4de88.mp4

---

I would recommend rebasing this PR and then taking a fresh look on how we can improve all this.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Using Toolbar is the right direction, but we should improve the UI based on different breakpoints even more. Let me know if I should take a deeper look into the Toolbar components.

**Comment by megan-hall** [comment]
@jerolimov - just providing my two cents based on your comments above. @TheRealJon keep me honest on my comments below based on our chat on Friday.
 
 ### 1. But couldn't we span and right align the last group (Wrap lines ... Expand) ??
_Agreed - we should keep wrap lines, raw, download and expand content functionality aligned on the right._ 

### 3. And then I'm personally also confused when reducing the width, even more, why do we have a filter and a more (3 dots) button which does the same?
_This is what PF recommends._

**Comment by TheRealJon** [comment]
I would disagree on your first point @jerolimov (right aligning the actions). We would need to override PatternFly to do this, which introduces unnecessary complexity. IMO, we should just align to the PatternFly Toolbar pattern. As far as point 3, there is a bug in PatternFly that is causing both the kebab and filter dropdowns to have the same content. I reported this to PF, but for now we could work around this by using a combination of the `ToolbarToggleGroup` and `OverflowMenu` components. See my last review about this: https://github.com/openshift/console/pull/11361#pullrequestreview-1010577267

**Comment by TheRealJon** [comment]
We also should include the search input in the ToolbarToggleGroup so that all of the toolbar items get collapsed at smaller widths. We can see this pattern in the PatternFly Toolbar component examples: https://www.patternfly.org/v4/components/toolbar/react/with-filters/

**Comment by jerolimov** [comment]
> We would need to override PatternFly to do this, which introduces unnecessary complexity. IMO, we should just align to the PatternFly Toolbar pattern.

PF provides an alignment prop for ToolbarGroup, which support breakpoints out of the box. So I would not call it “override PatternFly”, it's just using an existing feature explicit provided by their API.

See https://codesandbox.io/s/toolbar-example-4mx339?file=/index.js

**Comment by TheRealJon** [comment]
> ~I would disagree on your first point @jerolimov (right aligning the actions). We would need to override PatternFly to do this, which introduces unnecessary complexity. IMO, we should just align to the PatternFly Toolbar pattern.~

PF includes props that allow us to do this easily, so scratch that.

**Comment by TheRealJon** [comment]
I've been experimenting with the PF Toolbar today and testing edge cases for our proposed solution. It might be out of scope, but there is an issue with listing all of the log actions horizontally above the log viewer. The length of the log actions list is arbitrary. We provide an API (ConsoleExternalLogLink) that allows users to add an arbitrary number of links, each with arbitrary display text. This quickly causes issues with responsiveness, especially if the container name is long and the "Debug container" action is present.

I don't know that we need to worry about handling the case of multiple ConsoleExternalLogLinks, since that is likely a rare use case, but even one external link with relatively long display text will cause responsiveness issues near the 2xl breakpoint. Here are some gifs of my experimentation:

Near the 2XL (1450px) breakpoint, the labels start wrapping, and the list overflows:
screencast-mail google com-2022 06 22-14_05_19

Near the LG (992px) breakpoint, things look nice, although the search input shrinks a bit with longer container names:
screencast-mail google com-2022 06 22-14_07_31

**Comment by abhinandan13jan** [comment]
Closing this out as there are higher priority stories on HAC-DEV & HAC-BS side

**Pull request closed without merging**